File: backend/src/main/java/com/antshorttv/rbac/RbacController.java
package com.antshorttv.rbac;

import com.antshorttv.common.ApiResponse;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import java.util.List;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api")
public class RbacController {

    private final RbacService rbacService;

    public RbacController(RbacService rbacService) {
        this.rbacService = rbacService;
    }

    @GetMapping("/tenants/{tenantId}/roles")
    @RequirePermission("ROLE:VIEW")
    public ApiResponse<List<RoleResponse>> listRoles(@PathVariable Long tenantId) {
        return ApiResponse.success(rbacService.listRoles(tenantId));
    }

    @PostMapping("/tenants/{tenantId}/roles")
    @RequirePermission("ROLE:CREATE")
    public ApiResponse<RoleResponse> createRole(
        @PathVariable Long tenantId,
        @Valid @RequestBody CreateRoleRequest request,
        HttpServletRequest servletRequest
    ) {
        return ApiResponse.success(rbacService.createRole(tenantId, request, servletRequest));
    }

    @GetMapping("/tenants/{tenantId}/roles/{roleId}")
    @RequirePermission("ROLE:VIEW")
    public ApiResponse<RoleResponse> detail(@PathVariable Long tenantId, @PathVariable Long roleId) {
        return ApiResponse.success(rbacService.detail(tenantId, roleId));
    }

    @PutMapping("/tenants/{tenantId}/roles/{roleId}")
    @RequirePermission("ROLE:EDIT")
    public ApiResponse<RoleResponse> updateRole(
        @PathVariable Long tenantId,
        @PathVariable Long roleId,
        @Valid @RequestBody UpdateRoleRequest request,
        HttpServletRequest servletRequest
    ) {
        return ApiResponse.success(rbacService.updateRole(tenantId, roleId, request, servletRequest));
    }

    @DeleteMapping("/tenants/{tenantId}/roles/{roleId}")
    @RequirePermission("ROLE:DELETE")
    public ApiResponse<Void> deleteRole(
        @PathVariable Long tenantId,
        @PathVariable Long roleId,
        HttpServletRequest servletRequest
    ) {
        rbacService.deleteRole(tenantId, roleId, servletRequest);
        return ApiResponse.ok();
    }

    @PutMapping("/tenants/{tenantId}/roles/{roleId}/status")
    @RequirePermission("ROLE:EDIT")
    public ApiResponse<RoleResponse> updateStatus(
        @PathVariable Long tenantId,
        @PathVariable Long roleId,
        @Valid @RequestBody UpdateRoleStatusRequest request,
        HttpServletRequest servletRequest
    ) {
        return ApiResponse.success(rbacService.updateStatus(tenantId, roleId, request, servletRequest));
    }

    @GetMapping("/tenants/{tenantId}/roles/{roleId}/permissions")
    @RequirePermission("ROLE:VIEW")
    public ApiResponse<List<PermissionResponse>> rolePermissions(@PathVariable Long tenantId, @PathVariable Long roleId) {
        return ApiResponse.success(rbacService.rolePermissions(tenantId, roleId));
    }

    @PutMapping("/tenants/{tenantId}/roles/{roleId}/permissions")
    @RequirePermission("ROLE:EDIT")
    public ApiResponse<List<PermissionResponse>> updateRolePermissions(
        @PathVariable Long tenantId,
        @PathVariable Long roleId,
        @RequestBody UpdateRolePermissionsRequest request,
        HttpServletRequest servletRequest
    ) {
        return ApiResponse.success(rbacService.updateRolePermissions(tenantId, roleId, request, servletRequest));
    }

    @GetMapping("/tenants/{tenantId}/members/{memberId}/roles")
    @RequirePermission("MEMBER:VIEW")
    public ApiResponse<List<RoleResponse>> memberRoles(@PathVariable Long tenantId, @PathVariable Long memberId) {
        return ApiResponse.success(rbacService.memberRoles(tenantId, memberId));
    }

    @PutMapping("/tenants/{tenantId}/members/{memberId}/roles")
    @RequirePermission("MEMBER:INVITE")
    public ApiResponse<List<RoleResponse>> updateMemberRoles(
        @PathVariable Long tenantId,
        @PathVariable Long memberId,
        @RequestBody UpdateMemberRolesRequest request,
        HttpServletRequest servletRequest
    ) {
        return ApiResponse.success(rbacService.updateMemberRoles(tenantId, memberId, request, servletRequest));
    }

    @GetMapping("/permissions/tree")
    public ApiResponse<List<PermissionTreeNodeResponse>> permissionTree() {
        return ApiResponse.success(rbacService.permissionTree());
    }

}

File: frontend/src/services/ant-design-pro/rbacController.ts
// @ts-ignore
/* eslint-disable */
import { request } from "@umijs/max";

/** 此处后端没有提供注释 GET /api/permissions/tree */
export async function permissionTree(options?: { [key: string]: any }) {
  return request<API.ApiResponseListPermissionTreeNodeResponse>(
    "/api/permissions/tree",
    {
      method: "GET",
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /api/tenants/${param0}/members/${param1}/roles */
export async function memberRoles(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.memberRolesParams,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, memberId: param1, ...queryParams } = params;
  return request<API.ApiResponseListRoleResponse>(
    `/api/tenants/${param0}/members/${param1}/roles`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 PUT /api/tenants/${param0}/members/${param1}/roles */
export async function updateMemberRoles(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.updateMemberRolesParams,
  body: API.UpdateMemberRolesRequest,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, memberId: param1, ...queryParams } = params;
  return request<API.ApiResponseListRoleResponse>(
    `/api/tenants/${param0}/members/${param1}/roles`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      params: { ...queryParams },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /api/tenants/${param0}/roles */
export async function listRoles(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listRolesParams,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, ...queryParams } = params;
  return request<API.ApiResponseListRoleResponse>(
    `/api/tenants/${param0}/roles`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 POST /api/tenants/${param0}/roles */
export async function createRole(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.createRoleParams,
  body: API.CreateRoleRequest,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, ...queryParams } = params;
  return request<API.ApiResponseRoleResponse>(`/api/tenants/${param0}/roles`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    params: { ...queryParams },
    data: body,
    ...(options || {}),
  });
}

/** 此处后端没有提供注释 GET /api/tenants/${param0}/roles/${param1} */
export async function detail1(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.detail1Params,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, roleId: param1, ...queryParams } = params;
  return request<API.ApiResponseRoleResponse>(
    `/api/tenants/${param0}/roles/${param1}`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 PUT /api/tenants/${param0}/roles/${param1} */
export async function updateRole(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.updateRoleParams,
  body: API.UpdateRoleRequest,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, roleId: param1, ...queryParams } = params;
  return request<API.ApiResponseRoleResponse>(
    `/api/tenants/${param0}/roles/${param1}`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      params: { ...queryParams },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 DELETE /api/tenants/${param0}/roles/${param1} */
export async function deleteRole(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.deleteRoleParams,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, roleId: param1, ...queryParams } = params;
  return request<API.ApiResponseVoid>(
    `/api/tenants/${param0}/roles/${param1}`,
    {
      method: "DELETE",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 GET /api/tenants/${param0}/roles/${param1}/permissions */
export async function rolePermissions(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.rolePermissionsParams,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, roleId: param1, ...queryParams } = params;
  return request<API.ApiResponseListPermissionResponse>(
    `/api/tenants/${param0}/roles/${param1}/permissions`,
    {
      method: "GET",
      params: { ...queryParams },
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 PUT /api/tenants/${param0}/roles/${param1}/permissions */
export async function updateRolePermissions(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.updateRolePermissionsParams,
  body: API.UpdateRolePermissionsRequest,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, roleId: param1, ...queryParams } = params;
  return request<API.ApiResponseListPermissionResponse>(
    `/api/tenants/${param0}/roles/${param1}/permissions`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      params: { ...queryParams },
      data: body,
      ...(options || {}),
    }
  );
}

/** 此处后端没有提供注释 PUT /api/tenants/${param0}/roles/${param1}/status */
export async function updateStatus(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.updateStatusParams,
  body: API.UpdateRoleStatusRequest,
  options?: { [key: string]: any }
) {
  const { tenantId: param0, roleId: param1, ...queryParams } = params;
  return request<API.ApiResponseRoleResponse>(
    `/api/tenants/${param0}/roles/${param1}/status`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      params: { ...queryParams },
      data: body,
      ...(options || {}),
    }
  );
}
